# Lab session week 4: Analysis and cleaning

In this lab session you will learn how to work with analysing, cleaning and preprocessing data. You will use the kNN classifier in the end.

This weeks lab session consists of 5 small exercises:
1. Analysing your data
2. Imputation of missing values
3. Dealing with outliers
3. Preprocessing
4. Classification with kNN

In [12]:
from IPython.display import HTML
import matplotlib
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', 500)

## Exercise 1: analysing data with Pandas
During this exercise we will use the dataset `housedata.csv`. Please watch the following video to understand the describe function of Pandas:

In [13]:
HTML('<iframe width="560" height="315" src="https://www.youtube.com/embed/MaTAhxRsshE" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>')

c:\users\isahi\appdata\local\programs\python\python38\lib\site-packages\IPython\core\display.py:717: UserWarning: Consider using IPython.display.IFrame instead
  warnings.warn("Consider using IPython.display.IFrame instead")


This exercise will use a dataset that has been created using this dataset from Kaggle (https://www.kaggle.com/c/house-prices-advanced-regression-techniques/data). It might be useful to look up the column information for each of the columns in your own dataset before you proceed.

Now it is time to analyse our own dataset `housedata.csv`. Please do the following exercises:
1. Load your data into a dataframe
2. Have a look at the statistics using the describe function (hint: use include = 'all') and answer the following questions
    - Does your dataset have missing values?
    - Are their outliers?
    - How is the distribution of your data (hint: use plots)?

In [17]:
# Your code here

houses = pd.read_csv('housedata.csv')
houses.describe(include="all")

LotArea  LotFrontage BldgType  OverallQual  OverallCond  \
count     1460.000000  1201.000000     1378  1362.000000  1460.000000   
unique            NaN          NaN        5          NaN          NaN   
top               NaN          NaN     1Fam          NaN          NaN   
freq              NaN          NaN     1153          NaN          NaN   
mean     10516.828082    70.049958      NaN     6.095448     5.575342   
std       9981.264932    24.284752      NaN     1.386296     1.112799   
min       1300.000000    21.000000      NaN     1.000000     1.000000   
25%       7553.500000    59.000000      NaN     5.000000     5.000000   
50%       9478.500000    69.000000      NaN     6.000000     5.000000   
75%      11601.500000    80.000000      NaN     7.000000     6.000000   
max     215245.000000   313.000000      NaN    10.000000     9.000000   

          YearBuilt HeatingQC   Label  
count   1460.000000      1460    1460  
unique          NaN         5       3  
top             NaN        Ex  normal  
freq            NaN       741     593  
mean    1934.567808       NaN     NaN  
std      300.142482       NaN     NaN  
min     -491.000000       NaN     NaN  
25%     1952.000000       NaN     NaN  
50%     1972.000000       NaN     NaN  
75%     2000.000000       NaN     NaN  
max     2457.000000       NaN     NaN

**Your answers:**
- Does the data set have missing values? Which columns?
    - ... -> Yes, there is a difference in values in the "Count" row. If we take a look at the table, i would say the LotFrontage, BldgType and OverallQual columns have missing values because of the lower counts.
- Does the data set have outliers? Which columns?
    - ... -> yes there are surely outliers. if we look at the Yearbuilt column, you would expect that the yearbuilt would be in between 1900-2021 (now). But if we take a look at the min and max row of the Yearbuilt column we see significantly different values that exceeds or expectations. 2457 is in the future and -491 would be 491 b.c. i think? We could easily say there are outliers in this column.
- How is the distribution of your data?
    - ... -> there seems some not numerical data types in our data. it would be better to convert these types to numerical types for a ML algorithm. Bldgtype HealtingQc and Label are the NaN datatypes. Besides that we have a significant difference in numeric range. the LotArea is in between 1.300 and 215.245. while some other colums are in between 1 and 10. Data normalisation/standardization would be in handy for this dataset.

## Exercise 2: Imputation
We will now fix the problem of missing values by using imputation. Please watch the following video to understand how imputation works:

In [18]:
HTML('<iframe width="560" height="315" src="https://www.youtube.com/embed/D6saJ9R65L4" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>')

c:\users\isahi\appdata\local\programs\python\python38\lib\site-packages\IPython\core\display.py:717: UserWarning: Consider using IPython.display.IFrame instead
  warnings.warn("Consider using IPython.display.IFrame instead")


Lets try to do imputation on our own dataset.

1. Impute missing values (mean, median or mode) depending on the earlier found statistics
2. Use the Scikit-learn framework for imputation (http://scikit-learn.org/stable/modules/impute.html)

In [28]:
# Impute missing values for numerical data using the imputer
from sklearn.impute import SimpleImputer

# Your code here
houses.isnull()
houses.isnull().sum()
houses.isnull().sum()/len(houses)
#houses.drop('LotFrontage', inplace=True, axis=1)

imp = SimpleImputer(missing_values=np.nan, strategy='median')
houses['OverallQual'] = imp.fit_transform(houses['OverallQual'].values.reshape(-1,1))[:,0]

In [34]:
# Impute missing values for categorical values using pandas operations
houses.isnull().sum()

houses['BldgType'].value_counts()
houses['BldgType'].fillna('1Fam', inplace=True)
houses.isnull().sum()

# Your code here

LotArea        0
BldgType       0
OverallQual    0
OverallCond    0
YearBuilt      0
HeatingQC      0
Label          0
dtype: int64

## Exercise 3: Dealing with outliers
Since your dataset contains some outliers, we should fix this problem before proceeding to the next step. Read the following two (short) articles:
1. https://www.theanalysisfactor.com/outliers-to-drop-or-not-to-drop/
2. https://www.rapidinsightinc.com/handle-outliers/

Use the described techniques, find the outliers in your dataset and deal with them!

In [89]:
# Your code here
houses.describe(include="all")
houses.query('YearBuilt < 1800')
houses = houses[(houses.YearBuilt < 2021) & (houses.YearBuilt > 1800)]
houses.describe(include="all")

#houses.query('LotArea > 80000')
#houses['LotArea'].plot.hist()



LotArea BldgType  OverallQual  OverallCond    YearBuilt  \
count     1428.000000     1428  1428.000000  1428.000000  1428.000000   
unique            NaN        5          NaN          NaN          NaN   
top               NaN     1Fam          NaN          NaN          NaN   
freq              NaN     1213          NaN          NaN          NaN   
mean     10510.981092      NaN     6.084734     5.574930  1971.236695   
std      10028.036885      NaN     1.339065     1.117793    30.214811   
min       1300.000000      NaN     1.000000     1.000000  1872.000000   
25%       7538.750000      NaN     5.000000     5.000000  1954.000000   
50%       9468.500000      NaN     6.000000     5.000000  1972.500000   
75%      11601.500000      NaN     7.000000     6.000000  2000.000000   
max     215245.000000      NaN    10.000000     9.000000  2010.000000   

       HeatingQC   Label  
count       1428    1428  
unique         5       3  
top           Ex  normal  
freq         722     581  
mean         NaN     NaN  
std          NaN     NaN  
min          NaN     NaN  
25%          NaN     NaN  
50%          NaN     NaN  
75%          NaN     NaN  
max          NaN     NaN

## Exercise 4: Preprocessing
In order to use the data we have to preprocess the data. We have to perform several steps to make the data ready for classification.

1. Apply one-hot-encoding on categorical features (use Scikit-learn or getDummies in Pandas (see: http://queirozf.com/entries/one-hot-encoding-a-feature-on-a-pandas-dataframe-an-example))
2. Apply min-max-scaling or standardization (use Scikit-learn)

In [94]:
#1: Apply one hot encoding (using getdummies and concat?)

# Your code here
test = pd.concat([houses, pd.get_dummies(houses['BldgType'], prefix='BldgType')], axis=1)
test.drop(['BldgType'], axis=1, inplace=True)
test

LotArea  OverallQual  OverallCond  YearBuilt HeatingQC      Label  \
0        8450          7.0            5       2003        Ex     normal   
1        9600          6.0            8       1976        Ex     normal   
2       11250          7.0            5       2001        Ex  expensive   
3        9550          7.0            5       1915        Gd     normal   
4       14260          8.0            5       2000        Ex  expensive   
...       ...          ...          ...        ...       ...        ...   
1455     7917          6.0            5       1999        Ex     normal   
1456    13175          6.0            6       1978        TA     normal   
1457     9042          7.0            9       1941        Ex  expensive   
1458     9717          5.0            6       1950        Gd     normal   
1459     9937          5.0            6       1965        Gd     normal   

      BldgType_1Fam  BldgType_2fmCon  BldgType_Duplex  BldgType_Twnhs  \
0                 1                0                0               0   
1                 1                0                0               0   
2                 1                0                0               0   
3                 1                0                0               0   
4                 1                0                0               0   
...             ...              ...              ...             ...   
1455              1                0                0               0   
1456              1                0                0               0   
1457              1                0                0               0   
1458              1                0                0               0   
1459              1                0                0               0   

      BldgType_TwnhsE  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  
...               ...  
1455                0  
1456                0  
1457                0  
1458                0  
1459                0  

[1428 rows x 11 columns]

In [9]:
#2: Apply min-max scaling or standardization

# Your code here

## Exercise 5: Classification
1. Split the dataset in two parts (75%/25%) for training and testing.
2. Use the kNN algorithm from Scikit-learn and train a classifier
3. Validate using your testset
4. Play around with several hyperparameters to optimize your classification results

**Warning: this dataset will probably not give you great classification results (around 70% score)**

In [18]:
# Your code here